modules/gapi/include/opencv2/gapi/gkernel.hpp

```diff
@@ -141,87 +139,54 @@ namespace detail
         }
     };
 
-    // 4.1 - case for a single return value
-    // FIXME: How to avoid duplication here?
-    template<typename K, typename... Ins, typename Out>
-    struct MetaHelper<K, std::tuple<Ins...>, Out >
-    {
-        template<int... IIs>
-        static GMetaArgs getOutMeta_impl(const GMetaArgs &in_meta,
-                                         const GArgs &in_args,
-                                         detail::Seq<IIs...>)
-        {
-            // FIXME: decay?
-            using R = typename MetaType<Out>::type;
-            const R r = K::outMeta( get_in_meta<Ins>(in_meta, in_args, IIs)... );
-            return GMetaArgs{ GMetaArg(r) };
-        }
-        // FIXME: help users identify how outMeta must look like (via default impl w/static_assert?)
-
-        static GMetaArgs getOutMeta(const GMetaArgs &in_meta,
-                                    const GArgs &in_args)
-        {
-            return getOutMeta_impl(in_meta,
-                                   in_args,
-                                   typename detail::MkSeq<sizeof...(Ins)>::type());
-        }
-    };
-
     ////////////////////////////////////////////////////////////////////////////
     // Helper class to introduce tags to calls. By default there's no tag
     struct NoTag {
         static constexpr const char *tag() { return ""; }
     };
 
-} // namespace detail
+    template<typename, typename> class GKernelTypeImpl;
 
-// GKernelType and GKernelTypeM are base classes which implement typed ::on()
-// method based on kernel signature. GKernelTypeM stands for multiple-return-value kernels
-//
-// G_TYPED_KERNEL and G_TYPED_KERNEL_M macros inherit user classes from GKernelType and
-// GKernelTypeM respectively.
-
-template<typename K, typename... R, typename... Args>
-class GKernelTypeM<K, std::function<std::tuple<R...>(Args...)> >
-    : public detail::MetaHelper<K, std::tuple<Args...>, std::tuple<R...>>
-    , public detail::NoTag
-{
-    template<int... IIs>
-    static std::tuple<R...> yield(cv::GCall &call, detail::Seq<IIs...>)
+    template<typename K, typename... R, typename... Args>
+    class GKernelTypeImpl<K, std::function<std::tuple<R...>(Args...)> >
```

Comment:
self-review:
can partially specialize to accept:
```suggestion
    class GKernelTypeImpl<K, std::function<R(Args...)> >
```
seems like `R` (single-return-value) is good enough: can use `tuple_wrap_helper` where required for simple and correct behavior

Reply:
self-answer: not possible because we need to unpack single `R` into `R` instead of `tuple<R>` - this is done nicely in current approach